Implement pydeck with st.write

dcaminos opened the pull request:
Issue: https://github.com/streamlit/streamlit/issues/310

New Implementation:

```
import pandas as pd
import pydeck as pdk
import streamlit as st

DATA_URL = "https://raw.githubusercontent.com/uber-common/deck.gl-data/master/examples/geojson/vancouver-blocks.json"
LAND_COVER = [[[-123.0, 49.196], [-123.0, 49.324], [-123.306, 49.324], [-123.306, 49.196]]]

INITIAL_VIEW_STATE = pdk.ViewState(
  latitude=49.254,
  longitude=-123.13,
  zoom=11,
  max_zoom=16,
  pitch=45,
  bearing=0
)

polygon = pdk.Layer(
    'PolygonLayer',
    LAND_COVER,
    stroked=False,
    # processes the data as a flat longitude-latitude pair
    get_polygon='-',
    get_fill_color=[0, 0, 0, 20]
)

geojson = pdk.Layer(
    'GeoJsonLayer',
    DATA_URL,
    opacity=0.8,
    stroked=False,
    filled=True,
    extruded=True,
    wireframe=True,
    get_elevation='properties.valuePerSqm / 20',
    get_fill_color='[255, 255, properties.growth * 255]',
    get_line_color=[255, 255, 255],
    pickable=True
)

st.title("Pydeck GeoJsonLayer")
r = pdk.Deck(layers=[polygon, geojson],initial_view_state=INITIAL_VIEW_STATE)
st.write(r) # == st.deck_json_chart


icon_data = { "url":"https://img.icons8.com/plasticine/100/000000/marker.png", "width":128, "height":128, "anchorY": 128 }
data = pd.read_json("https://raw.githubusercontent.com/uber-common/deck.gl-data/master/website/bart-stations.json")
data['icon_data']=None
for i in data.index:
     data['icon_data'][i]=icon_data

view_state = pdk.ViewState(
    longitude=-122.22,
    latitude=37.76,
    zoom=9,
    pitch=50
)

icon_layer2 = pdk.Layer(
    type='IconLayer',
    data=data,
    get_icon='icon_data',
    get_size=4,
    pickable= True,
    size_scale= 15,
    get_position= 'coordinates'
)

st.title("Pydeck IconLayer")
r2 = pdk.Deck(layers=[icon_layer2],initial_view_state=view_state)
st.deck_json_chart(r2) #== st.write
```

<img width="713" alt="Screen Shot 2019-11-07 at 11 52 58 AM" src="https://user-images.githubusercontent.com/22157341/68401144-fc021980-0157-11ea-8a9b-f73136f75097.png">

treuille commented:
## Summary

@dcaminos: Can you please answer question 1 and give you thoughts on the feasibility of suggestions 1 and 2.

@tvst : Can you please sign off on suggestions 1 and 2?

## Question 1

Will this also solve #522 and #311?

## Suggestion 1

Can we also enable the following?

```python
r = pdk.Deck(layers=[hexagon_layer], initial_view_state=view_state)
st.pydeck(r)
```

And also (equivalently) the following?

```python
r = pdk.Deck(layers=[hexagon_layer], initial_view_state=view_state)
st.write(r)
```

## Suggestion 2:

Also I think

```python
st.deck_gl_chart(json=r.to_json())
```

is an implementation detail and shouldn't be added in the documentation. In fact, in the future `st.deck_gl_chart` should output a deprecation warning.

dcaminos commented:
### Sumary
Pydeck doc: https://deckgl.readthedocs.io/en/latest/index.html

### Question 1
[522](https://github.com/streamlit/streamlit/issues/522l): No right now, because I only did the JSON solution (some parameters like 'get_icon' need be a function). I can solve this in the future but I had no time yet. 
[311](https://github.com/streamlit/streamlit/issues/311): Yes, in fact I added an example about GeoJsonLayer [code](https://github.com/dcaminos/streamlit/blob/issue310/deck_gl_chart_use_pydeck/examples/vancouver_property_values.py)
<img width="714" alt="Screen Shot 2019-10-31 at 4 49 53 PM" src="https://user-images.githubusercontent.com/22157341/67981039-85689780-fbfe-11e9-9963-9068b7f09476.png">

### Suggestion 1
Yes for both: 
```
r = pdk.Deck(...)
st.pydeck(r)
st.write(r)
```
But like I started with the json solution, this is not implemented yet. I wanna do it in another PR.

Suggestion 2### 
Yes, this is implemented on this PR.

I will add the st.deck_gl_chart deprecation warning on next PR (when implement the st.write(r))

dcaminos commented:
### New requirement

FIRST PR:
**st.write( pydeck_obj )** ←should show a DeckGL chart!
**st.write( json )** ← should do the same thing it does today: show raw JSON
**st.deck_gl_chart()**←should still work like today

NEXT PR:
**st.write( pydeck_obj )** ←should still work like in the 1st PR
**st.write( json )** ← should do the same thing it does today: show raw JSON
**st.deck_gl_chart( … )** ←deprecated
Show warning saying "deprecated. will be removed in 3 months. use pydeck."
But also allow it to work
**st.pydeck( … )** ←pass JSON or pydeck_obj

kuttacondition commented:
Look forward trying out the geojson support in a future streamlit release.

dcaminos commented:
> ## Summary
> @dcaminos: Can you please answer question 1 and give you thoughts on the feasibility of suggestions 1 and 2.
> 
> @tvst : Can you please sign off on suggestions 1 and 2?
> 
> ## Question 1
> Will this also solve #522 and #311?
> 
> ## Suggestion 1
> Can we also enable the following?
> 
> ```python
> r = pdk.Deck(...)
> st.pydeck(r)
> ```
> 
> And also (equivalently) the following?
> 
> ```python
> r = pdk.Deck(...)
> st.write(r)
> ```
> 
> ## Suggestion 2:
> Also I think
> 
> ```python
> st.deck_gl_chart(json=r.to_json())
> ```
> 
> is an implementation detail and shouldn't be added in the documentation. In fact, in the future `st.deck_gl_chart` should output a deprecation warning.

@treuille 

### Question 1:
Yes, I added an example at the top of this PR.

### Suggestion 1:
Yes, on this PR I added support for `st.write(pydeck_obj)` and `st.deck_json_chart(pydeck_obj)`
We can easily change the name `deck_json_chart` if you want.

Suggestion 2:
With @tvst agree do that in next PR, but should be very very easy to do.

PS: I added a deprecation warning for `st.deck_gl_chart`

treuille commented:
OMG @dcaminos. This looks _amazing_!!

@tvst : We need to update the uber demo which currently [uses the old api](https://github.com/streamlit/demo-uber-nyc-pickups/blob/master/app.py#L55)? How should we track that?

tvst commented on frontend/src/components/core/Block/Block.tsx:
```diff
@@ -41,6 +41,9 @@ const DataFrame = React.lazy(() => import("components/elements/DataFrame/"))
   import("components/elements/DeckGlChart/")
 )
+const DeckJsonChart = React.lazy(() =>
+  import("components/elements/DeckJsonChart/")
```
I think `PyDeckChart` would be a clearer name.

Same in the proto, etc.

tvst commented on frontend/src/components/elements/DeckGlChart/DeckGlChart.tsx:
```diff
@@ -96,33 +97,73 @@ class DeckGlChart extends React.PureComponent<PropsWithHeight, State> {
+          style={{ width: this.initialViewState.width }}
+        >
+          <div className="markdown-text-container">
+            <StreamlitMarkdown
```
Instead of adding this warning via JS, it's best to do in Python inside `DeltaGenerator.deck_gl_chart`:

```py
def deck_gl_chart(...):
  st.warning("""
    WARNING GOES HERE
    """)
  # rest of the code goes here
```

Then we can also easily add a keyword parameter that turns the warning off (which is something we should totally do!):

```py
def deck_gl_chart(..., suppress_deprecation_warning=False):
  if not suppress_deprecation_warning:
    st.warning("""
      WARNING GOES HERE
      """)
  # rest of the code goes here
```

(then you can also undo all changes to this file)

tvst commented on frontend/src/components/elements/DeckJsonChart/DeckJsonChart.tsx:
```diff
@@ -0,0 +1,126 @@
+
+    return (
+      <div
+        className="deckglchart stDeckJsonChart"
```
Why do we need the class `deckglchart`?

tvst commented on frontend/src/components/elements/DeckJsonChart/DeckJsonChart.scss:
```diff
@@ -0,0 +1,19 @@
+ */
+
+.deckglchart {
+  position: relative;
```
Why is this needed?

tvst commented on lib/Pipfile:
```diff
@@ -68,3 +69,6 @@ numpy = "*"
 numpy = "<1.17.0"
+
+[pipenv]
+allow_prereleases = true
```
Why do we need this? 

I think this will make all `*` versions possibly be pre-release. Sounds like something we should avoid, if at all possible

tvst commented on lib/Pipfile:
```diff
@@ -58,6 +58,7 @@ tzlocal = "*"
 validators = "*"
 watchdog = "*"
 # Only Python 3.x packages
+pydeck = "*"
```
I think the comment above this doesn't make sense where it is. But it's also a useless comment anyway, so please remove it.

Same with the python 2.x comment below.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -39,6 +39,7 @@
 from streamlit.proto import ForwardMsg_pb2
 from streamlit.proto import Text_pb2
 from streamlit import get_report_ctx
+import streamlit.elements.map as streamlit_map
```
Move this `import` inside the methods that use it.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+
+        This API closely follows Deck.GL's JavaScript API
+        (https://deck.gl/#/documentation), with a few small adaptations and
+        some syntax sugar.
```
This paragraph doesn't apply to this method.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
         deck_gl.marshall(element.deck_gl_chart, spec, **kwargs)
 
+    @_with_element
+    def deck_json_chart(self, element, pydeck_obj=None):
```
Rename to `pydeck_chart`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+        ...    get_size=4,
+        ...    pickable= True,
+        ...    size_scale= 15,
+        ...    get_position= 'coordinates')
```
Remove space after `=`. Same above.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+        ...    size_scale= 15,
+        ...    get_position= 'coordinates')
+
+        >>> r2 = pdk.Deck(layers=[icon_layer2],initial_view_state=view_state)
```
Add space after `,`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+        >>> data = pd.read_json("https://raw.githubusercontent.com/uber-common/deck.gl-data/master/website/bart-stations.json")
+        >>> data['icon_data']=None
+        >>> for i in data.index:
+        >>>      data['icon_data'][i]=icon_data
```
Should be `...`, not `>>>`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+
+        >>> icon_data = { "url":"https://img.icons8.com/plasticine/100/000000/marker.png", "width":128, "height":128, "anchorY": 128 }
+        >>> data = pd.read_json("https://raw.githubusercontent.com/uber-common/deck.gl-data/master/website/bart-stations.json")
+        >>> data['icon_data']=None
```
Add space around `=`

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+        >>> r = pdk.Deck(layers=[polygon, geojson],initial_view_state=INITIAL_VIEW_STATE)
+        >>> st.write(r)
+
+        >>> icon_data = { "url":"https://img.icons8.com/plasticine/100/000000/marker.png", "width":128, "height":128, "anchorY": 128 }
```
Line too long. Break this into multiple lines.

Same below.

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2422,6 +2421,98 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+
+        >>> r = pdk.Deck(layers=[polygon, geojson],initial_view_state=INITIAL_VIEW_STATE)
+        >>> st.write(r)
+
```
This is the start of a whole new example, right? So add some text here to let the user know:

```
        >>> r = pdk.Deck(layers=[polygon, geojson],initial_view_state=INITIAL_VIEW_STATE)
        >>> st.write(r)

        And here is an example of placing icons on a map with an `IconLayer`:

        >>> icon_data = ...
```

tvst commented on lib/streamlit/elements/map.py:
```diff
@@ -13,14 +13,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+"""A Python wrapper around DeckGl."""
```
Add an empty line below this.

Also, this pydoc isn't 100% correct. This would be more correct: `A wrapper for simple PyDeck scatter charts`

tvst commented on lib/streamlit/elements/map.py:
```diff
@@ -13,14 +13,12 @@
+import json
 
+DEFAULT_MAP = '{"initialViewState":{"latitude":0,"longitude":0,"pitch":0,"zoom":1},"mapStyle":["mapbox://styles/mapbox/light-v10"]}'
+DEFAULT_COLOR = [200, 30, 0, 160]
```
Please prefix with `_` (underscore) all private members of this module, like `_DEFAULT_COLOR`

tvst commented on lib/streamlit/elements/map.py:
```diff
@@ -64,19 +61,12 @@ def _get_zoom_level(distance):
-
-def marshall(element, data, zoom=None):
-    """Marshall a proto with DeckGL chart info.
+            return i
```
Why change to `i`?

dcaminos replied:
Based on this [documentation](https://github.com/uber/deck.gl/blob/master/docs/api-reference/map-view.md) the min zoom is 0 and the max is 20.

And with the +1 the view had a lot padding.

tvst commented on lib/streamlit/elements/map.py:
```diff
@@ -13,14 +13,12 @@
-LOGGER = get_logger(__name__)
+import json
 
+DEFAULT_MAP = '{"initialViewState":{"latitude":0,"longitude":0,"pitch":0,"zoom":1},"mapStyle":["mapbox://styles/mapbox/light-v10"]}'
```
I think it makes more sense to keep `DEFAULT_MAP` as a Python `dict` rather than a JSON-formatted `str`.

tvst commented on lib/streamlit/elements/map.py:
```diff
@@ -64,19 +61,12 @@ def _get_zoom_level(distance):
-    This is a shorthand for DeltaGenerator.deck_gl_chart,
-    which will auto center and auto zoom the chart.
-    If it is needed you can specify the zoom param.
+def embed_data(data, zoom):
```
A clearer name would be `to_deckgl_json`

tvst commented on lib/tests/streamlit/streamlit_test.py:
```diff
@@ -428,7 +428,7 @@ def test_st_pyplot(self):
         # Manually calculated by letting the test fail and copying and
         # pasting the result.
-        checksum = "DTuIkOADCFAAEAmPL/AFE92BIZHj8WAAAAAElFTkSuQmCC"
+        checksum = "n1ZpaandJQILIkAAAJ7EQ3QAgCkECADAlP8PvwXGkVVfFJkAAAAASUVORK5CYII="
```
Why is this changed? You didn't touch pyplot in you code, so the test shouldn't fail unless something is actually broken.

tvst commented on proto/streamlit/proto/Element.proto:
```diff
@@ -77,7 +78,7 @@ message Element {
-
-    // Next ID: 29
+    DeckJsonChart deck_json_chart = 30;
+    // Next ID: 31
```
FYI: this PR is going to have a merge conflict when #506 lands

tvst commented on lib/streamlit/util.py:
```diff
@@ -472,6 +472,9 @@ def is_namedtuple(x):
 
+def is_pydeck(obj):
+    """True if input looks like a pydeck chart."""
+    return is_type(obj, "pydeck.bindings.deck.Deck")
```
These `is_foo` methods have moved to a file called `type_util.py`. Please merge the latest `develop` and move this one there too.

dcaminos commented:
@tvst I think pydeck changed her implementation, get_position only support string type.
https://deckgl.readthedocs.io/en/latest/layer.html#understanding-get-position

tvst commented:
- Rename some things and improve docs:
  https://github.com/streamlit/streamlit/commit/11d219b0405953ee9f0f7b00cd6d6b85fdbbebe3
  (this is based on an old commit! Sorry!)

- Also this:
  https://github.com/streamlit/streamlit/commit/4095dd24c0723a02b664af77c4731c8ef6a95da5
  (this is based on the latest commit)

- Make sure st.map matches today's behavior. (Make sure the map's "pitch"
  is the same as it used to be. I.e. the angle the camera.)

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2557,10 +2535,89 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+        using Deck.GL' JSON converter JavaScript API
+        (https://github.com/uber/deck.gl/tree/master/modules/json)
+
+
```
Only 1 blank line

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2557,10 +2535,89 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+           height: 530px
+
+        """
+
```
Remove blank line

tvst commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -2557,10 +2535,89 @@ def deck_gl_chart(self, element, spec=None, **kwargs):
+
+        """
+
+        if pydeck_obj == None:
```
use `is None`

tvst commented on lib/streamlit/config.py:
```diff
@@ -282,6 +282,14 @@ def _global_unit_test():
 )
 
+_create_option(
+    "global.suppressDeprecationWarning",
```
add `s` at the end: `suppressDeprecationWarnings`

tvst commented on frontend/src/components/elements/DeckGlJsonChart/DeckGlJsonChart.tsx:
```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2018-2019 Streamlit Inc.
```
Please update all copyrights to 2020

tvst commented on frontend/src/components/elements/DeckGlJsonChart/DeckGlJsonChart.tsx:
```diff
@@ -0,0 +1,125 @@
+import Immutable from "immutable"
+import { StaticMap } from "react-map-gl"
+import FullScreenWrapper from "components/shared/FullScreenWrapper"
+import "mapbox-gl/dist/mapbox-gl.css"
```
Don't you need to add this scss file?
https://github.com/streamlit/streamlit/commit/11d219b0405953ee9f0f7b00cd6d6b85fdbbebe3#diff-028a1ac45d1cb5284bc51fdc0faf32ad

dcaminos replied:
No. is not necessary
